src/magnifier.c

```diff
@@ -0,0 +1,296 @@
+	static struct wlr_buffer *tmp_buffer = NULL;
+	static struct wlr_texture *tmp_texture = NULL;
+
+	/* TODO: This looks way too complicated to just get the used format */
+	struct wlr_drm_format wlr_drm_format = {0};
+	struct wlr_shm_attributes shm_attribs = {0};
+	struct wlr_dmabuf_attributes dma_attribs = {0};
+	if (wlr_buffer_get_dmabuf(output_buffer, &dma_attribs)) {
+		wlr_drm_format.format = dma_attribs.format;
+		wlr_drm_format.len = 1;
+		wlr_drm_format.modifiers = &dma_attribs.modifier;
+	} else if (wlr_buffer_get_shm(output_buffer, &shm_attribs)) {
+		wlr_drm_format.format = shm_attribs.format;
+	} else {
+		wlr_log(WLR_ERROR, "Failed to read buffer format");
+		return;
+	}
+
+	/* Fetch scale-adjusted cursor coordinates */
+	struct server *server = output->server;
+	struct theme *theme = server->theme;
+	struct wlr_cursor *cursor = server->seat.cursor;
+	double ox = cursor->x;
+	double oy = cursor->y;
+	wlr_output_layout_output_coords(server->output_layout, output->wlr_output, &ox, &oy);
+	ox *= output->wlr_output->scale;
+	oy *= output->wlr_output->scale;
+	if (theme->mag_width == -1 || theme->mag_height == -1) {
+		fullscreen = true;
+	}
+	if ((ox < 0 || oy < 0 || ox >= output_buffer->width || oy >= output_buffer->height)
+		&& fullscreen) {
+		return;
+	}
+
+	if (mag_scale == 0.0) {
+		mag_scale = theme->mag_scale;
+	}
+	if (mag_increment == 0.0) {
+		mag_increment = theme->mag_increment;
+	}
+
+	if (fullscreen) {
+		width = output_buffer->width;
+		height = output_buffer->height;
+		x = 0;
+		y = 0;
+	} else {
+		width = theme->mag_width + 1;
+		height = theme->mag_height + 1;
+		x = ox - (theme->mag_width / 2.0);
+		y = oy - (theme->mag_height / 2.0);
+	}
+	double cropped_width = width;
+	double cropped_height = height;
+	double dst_x = 0;
+	double dst_y = 0;
+
+	/* Ensure everything is kept within output boundaries */
+	if (x < 0) {
+		cropped_width += x;
+		dst_x = x * -1;
+		x = 0;
+	}
+	if (y < 0) {
+		cropped_height += y;
+		dst_y = y * -1;
+		y = 0;
+	}
+	cropped_width = MIN(cropped_width, (double)output_buffer->width - x);
+	cropped_height = MIN(cropped_height, (double)output_buffer->height - y);
+
+	/* (Re)create the temporary buffer if required */
+	if (tmp_buffer && (tmp_buffer->width != width || tmp_buffer->height != height)) {
+		wlr_log(WLR_ERROR, "tmp buffer size changed, dropping");
+		assert(tmp_texture);
+		wlr_texture_destroy(tmp_texture);
+		wlr_buffer_drop(tmp_buffer);
+		tmp_buffer = NULL;
+		tmp_texture = NULL;
+	}
+	if (!tmp_buffer) {
+		tmp_buffer = wlr_allocator_create_buffer(
+			server->allocator, width, height, &wlr_drm_format);
+	}
+	if (!tmp_buffer) {
+		wlr_log(WLR_ERROR, "Failed to allocate temporary magnifier buffer");
+		return;
+	}
+
+	/* Extract source region into temporary buffer */
+
+	struct wlr_render_pass *tmp_render_pass = wlr_renderer_begin_buffer_pass(
+		server->renderer, tmp_buffer, NULL);
+
+	/* FIXME, try to re-use the existing output texture instead */
```

Comment:
I don't think we can do this so suggest removing the comment.

Reply:
Done